core/src/main/scala/sttp/client/testing/SttpBackendStub.scala

```diff
@@ -211,3 +288,109 @@ object SttpBackendStub {
     }
   }
 }
+
+/**
+  * Websockets use can differ, for akka backend crucial part is the handler.
+  * For other backends sttp provides handlers that return [[WebSocket]]
+  * instance to send and receive messages.
+  */
+private[testing] sealed trait WhenOpenWebsocket[F[_], +WS_HANDLER[_]]
+
+private[testing] case class UseHandler[F[_], WS_RESULT, WS_HANDLER[_]](
+    headers: Headers,
+    useHandler: Any => WS_RESULT
+) extends WhenOpenWebsocket[F, WS_HANDLER]
+
+private[testing] case class ReturnWebsocketResponse[F[_], WS_RESULT](headers: Headers, response: F[WS_RESULT])
+    extends WhenOpenWebsocket[F, NothingT]
+
+/**
+  * A simple stub for web sockets that uses a queue of events for receive.
+  * New messages can be added to queue when `send` is invoked.
+  * For more complex cases, please provide own implementation of [[WebSocket]].
+  */
+class WebSocketStub[S](
+    initialResponses: List[Try[Either[WebSocketEvent.Close, WebSocketFrame.Incoming]]],
+    initialState: S,
+    makeNewEvents: (S, WebSocketFrame) => (S, List[Try[Either[WebSocketEvent.Close, WebSocketFrame.Incoming]]])
+) {
+
+  /** Returns stub that has same initial messages but replaces funtion that adds messages to receive when `sent` is called. */
+  def thenRespond(addReceived: WebSocketFrame => List[WebSocketFrame.Incoming]): WebSocketStub[Unit] =
+    thenRespondWith(
+      addReceived.andThen(_.map(m => Success(Right(m): Either[WebSocketEvent.Close, WebSocketFrame.Incoming])))
+    )
+
+  /** More powerful version of [[thenRespond()]]. Allows to use function that calls WebSocket to close or fails. */
+  def thenRespondWith(
+      addReceived: WebSocketFrame => List[Try[Either[WebSocketEvent.Close, WebSocketFrame.Incoming]]]
```

Comment:
maybe this should be: `Try[Either[Close, List[Incoming]]]`?

Reply:
Do you mean 1 to 1 or just using other imports to shorten?
We don't offer full flexibility in stub, but I think having possibility to add 0 to many new responses for on 1 input message for price of having `List` in this signature is a good tradeoff.

Reply:
I mean reordering the effects :) The outer effect - Try - as everything can fail. Then, we can either close the websocket or receive 0-n messages. There's no sense receiving messages after a close, for example :)

Reply:
(and the stub should work like that - once a `Close` has been received by the client, the websocket (here: stub) becomes closed, and any further `.receive` attempts end with an exception)

Reply:
done